bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/DefaultModelResourceLoader.java

```diff
@@ -0,0 +1,336 @@
+      Command decoded = commandCodec.decode(getEditingDomain(getResourceSet(modeluri)), command);
+      getEditingDomain(getResourceSet(modeluri)).execute(decoded);
+   }
+
+   @Override
+   public Command undo(final String modeluri) {
+      ModelServerEditingDomain editingDomain = getEditingDomain(getResourceSet(modeluri));
+      Command undoCommand = editingDomain.getCommandStack().getUndoCommand();
+      if (editingDomain.canUndo()) {
+         editingDomain.undo();
+         return createInverseCommand(undoCommand);
+      }
+      return null;
+   }
+
+   @SuppressWarnings("checkstyle:CyclomaticComplexity")
+   private Command createInverseCommand(final Command undoCommand) {
+      if (undoCommand instanceof CompoundCommand) {
+         CompoundCommand undoCompoundCommand = (CompoundCommand) undoCommand;
+
+         CompoundCommand inverseCompoundCommand = new CompoundCommand();
+         for (Command next : undoCompoundCommand.getCommandList()) {
+            inverseCompoundCommand.append(createInverseCommand(next));
+         }
+         return inverseCompoundCommand;
+
+      } else if (undoCommand instanceof AddCommand) {
+         AddCommand undoAddCommand = (AddCommand) undoCommand;
+         return RemoveCommand.create(undoAddCommand.getDomain(), undoAddCommand.getOwner(),
+            undoAddCommand.getFeature(), undoAddCommand.getResult());
+
+      } else if (undoCommand instanceof RemoveCommand) {
+         RemoveCommand undoRemoveCommand = (RemoveCommand) undoCommand;
+         return AddCommand.create(undoRemoveCommand.getDomain(), undoRemoveCommand.getOwner(),
+            undoRemoveCommand.getFeature(), undoRemoveCommand.getResult());
+      } else if (undoCommand instanceof SetCommand) {
+         // FIXME: Handle the UNSET value, see also DefaultCommandCodec where it is also not yet implemented
+         SetCommand undoSetCommand = (SetCommand) undoCommand;
+         return SetCommand.create(undoSetCommand.getDomain(), undoSetCommand.getOwner(),
+            undoSetCommand.getFeature(), undoSetCommand.getOldValue(), undoSetCommand.getIndex());
+      } else if (undoCommand instanceof ReplaceCommand) {
+         // TODO see also DefaultCommandCodec
+      } else if (undoCommand instanceof MoveCommand) {
+         // TODO see also DefaultCommandCodec
+      }
+      return undoCommand;
+   }
+
+   @Override
+   public Command redo(final String modeluri) {
+      ModelServerEditingDomain editingDomain = getEditingDomain(getResourceSet(modeluri));
+      Command redoCommand = editingDomain.getCommandStack().getRedoCommand();
+      if (editingDomain.canRedo()) {
+         editingDomain.redo();
+         return redoCommand;
+      }
+      return null;
+   }
+
+   @Override
+   public boolean save(final String modeluri) {
+      ResourceSet resourceSet = getResourceSet(modeluri);
+      boolean result = resourceSet.getResources().stream().allMatch(this::saveResource);
```

Comment:
This actually ignores the 'modelUri', see also [1]. I suggest to make the model uri optional and only save the single resource if it is given and all resources otherwise. Or do you think that saving only a single resource may cause problems down the road? Because currently we get the editing domain for the resource and call 'saveIsDone' on it.

[1] https://spectrum.chat/emfcloud/modelserver/save-api-ignores-parameter-modeluri~a9b386f6-80d2-4fbf-a8f9-5fbc354b6a6e

Reply:
Yes that's true, as we discussed, I now introduced a `saveAll` which saves all `Resource`s of the available `ResourceSet`s. The `save(modelUri)` is fixed now and only saves the dedicated `Resource` behind the `modelUri`.